### PR TITLE
S3 first byte latency metric

### PR DESCRIPTION
**Pull request by JVerwolf**
This PR adds a metrics for s3 backed blob store first byte latencies.

### CODE REVIEW

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java [hunk 1]
```diff
@@ -172,6 +177,7 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
                 assert assertConsistencyBetweenHttpRequestAndOperation(request, operation);
                 counter.add(getRequestCount(request));
                 requestCounter.incrementBy(getRequestCount(request), attributes);
+                httpRequestTimeInMicrosCounter.incrementBy(getHttpRequestTimeInMicros(request), attributes);
```
Comment (JVerwolf) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:180]: Similar to the APM metric above, this should multiplex to the various use-cases, as per the `attributes` map.

**Comment by JVerwolf** [comment]
I'm note sure if this should be part of this PR, but I could also refactor the metrics here, including the preexisting counter, to be registered as part of the construction of a new `S3RepositoryAPMMetrics` class, which would then be a dependency of the `S3Repository` class (e.g. see the `BlobCacheMetrics` class for the pattern).  This keeps metric registration out of the class construction.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java [hunk 2]
```diff
@@ -112,7 +116,8 @@ class S3BlobStore implements BlobStore {
         this.threadPool = threadPool;
         this.snapshotExecutor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
         this.meterRegistry = meterRegistry;
-        this.requestCounter = this.meterRegistry.getLongCounter(METRIC_REQUESTS_COUNT);
+        this.requestCounter = this.meterRegistry.getLongCounter(METRIC_REQUESTS_COUNT); // here
```
Comment (piergm) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:119]: Nit: remove the `// here` comment

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java [hunk 3]
```diff
@@ -48,7 +49,11 @@ public RepositoriesModule(
         RecoverySettings recoverySettings,
         TelemetryProvider telemetryProvider
     ) {
+        // TODO: refactor APM metrics into their own class, passed in as a dependancy.
```
Comment (piergm) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java:52]: I do agree with you to refactor this out into their own class, but I'd rather do it in a separate PR.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java [hunk 4]
```diff
@@ -36,6 +36,7 @@
 public final class RepositoriesModule {
 
     public static final String METRIC_REQUESTS_COUNT = "repositories.requests.count";
+    public static final String HTTP_REQUEST_TIME_IN_MICROS_COUNT = "repositories.requests.httpRequestTimeInMicros.count";
```
Comment (piergm) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java:39]: In the "Metrics Naming Guideline Proposal" doc it says not to include units in the metric name but instead put them in the unit and to use snake_case instead of camelCase. So I would rename the metric to something like:
`repositories.requests.http_request_time.count` and keep the `micros` as unit.

**Review by piergm — APPROVED** [review]
LGTM, I left some small comments

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java (hunk 3 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java:52]: Yeah, it may involve changing a bunch of the tests, cluttering these changes up. Agreed.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java [hunk 5]
```diff
@@ -206,6 +213,32 @@ private static long getRequestCount(Request<?> request) {
...
     }
 
+    /**
+     * Used for APM style metrics to measure statics about performance. This is not for billing.
+     */
+    private static long getHttpRequestTimeInMicros(Request<?> request) {
+        List<TimingInfo> requestTimesIncludingRetries;
+        requestTimesIncludingRetries = request.getAWSRequestMetrics()
+            .getTimingInfo()
+            .getAllSubMeasurements(AWSRequestMetrics.Field.HttpRequestTime.name());
```
Comment (ywangd) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:223]: `HttpRequestTime` including time spent on acquiring an existing or creating a new connection, marshalling the input etc. Do we want to include for measuring "time to first byte"?

IIUC, last time we discussed about it, we prefer to exclude time spent on local node and measure only for networking, e.g. something like the sum of `HttpClientSendRequestTime` and `HttpClientReceiveResponseTime`. Maybe the preference has since changed?

**Review by ywangd — COMMENTED** [review]
I have a question regarding how we want to measure "time to first byte". Also, do we need any test for the change?

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java (hunk 3 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java]: FYI, I have an ongoing PR https://github.com/elastic/elasticsearch/pull/102505 that adds more this registration. I can take up the task to refactor them after both PRs are done.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java (hunk 5 above)
Comment (JVerwolf) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:223]: Thanks for the well-considered comment @ywangd !

@original-brownbear I think this one's for you.  We discussed using `TTFB` (in addition to another metric that measures the total time to fetch the block) as a way to infer what portion of the latency is  on us vs. what's on AWS.  I do see @ywangd 's point regarding the measurement of the total networking time.  

I suppose that these submeasurements compose an overlapping Venn diagram containing `Our latency` + ( `Our latency` + `AWS latency`) + `AWS latency`. If we want to know why things are slow (us vs AWS), the tricky bit is getting signal given the fact that there's overlap in responsibility for a number of these  (`Our latency` + `AWS latency`).  

Seeing as you've got a better background+understanding of where the overlap exists, WDYT?


From the docs:
 [HttpRequestTime](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/util/AWSRequestMetrics.Field.html#HttpRequestTime)
> Number of milliseconds taken to acquire a connection from the connection pool (or create a new connection), send the request (including reading the user's input stream for streaming uploads), and receive the response status and headers. This does not include the time needed to download the response payload from the AWS service.

[HttpClientSendRequestTime](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/util/AWSRequestMetrics.Field.html#HttpClientSendRequestTime)
> Number of milliseconds taken to write the request headers and payload to the underlying socket that is connected to AWS. This does not include any time spent acquiring a connection from the connection pool (or creating a new connection).

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java (hunk 3 above)
Comment (JVerwolf) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java]: Thanks @ywangd 🙂

**Review by original-brownbear — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java [hunk 6]
```diff
@@ -66,6 +67,13 @@ public RepositoriesModule(
             .registerLongHistogram(METRIC_EXCEPTIONS_HISTOGRAM, "repository request exception histogram", "unit");
         telemetryProvider.getMeterRegistry()
             .registerLongHistogram(METRIC_THROTTLES_HISTOGRAM, "repository request throttle histogram", "unit");
+        telemetryProvider.getMeterRegistry()
+            .registerLongHistogram(
+                HTTP_REQUEST_TIME_IN_MICROS_HISTOGRAM,
+                "HttpRequestTime in microseconds expressed as as a histogram",
+                "micros"
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java:74]: Should it be microseconds rather then micros?

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ public final class RepositoriesModule {
     public static final String METRIC_UNSUCCESSFUL_OPERATIONS_COUNT = "es.repositories.operations.unsuccessful.count";
     public static final String METRIC_EXCEPTIONS_HISTOGRAM = "es.repositories.exceptions.histogram";
     public static final String METRIC_THROTTLES_HISTOGRAM = "es.repositories.throttles.histogram";
-
+    public static final String HTTP_REQUEST_TIME_IN_MICROS_HISTOGRAM = "es.repositories.requests.http_request_time.histogram";
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesModule.java:45]: I wonder if we need to repeat request twice? I think this could be shortened to:
`es.repositories.requests.duration.histogram`
